### PR TITLE
Add test to check for corrupted StringBuilder state after OutOfMemoryException

### PR DESCRIPTION
**Pull request by jkotas**
Regression test for dotnet/coreclr#20718 and dotnet/coreclr#20719

### CODE REVIEW

**Comment by jkotas** [comment]
CoreCLR PR: https://github.com/dotnet/coreclr/pull/20724

File: src/System.Runtime/tests/System/Text/StringBuilderTests.netcoreapp.cs [hunk 1]
```diff
@@ -445,5 +445,51 @@ public static void EqualsIgnoresMaxCapacity()
 
             Assert.True(sb1.Equals(sb2));
         }
+
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsRobustUnderMemoryPressure))]
```
Comment (danmosemsft) [inline src/System.Runtime/tests/System/Text/StringBuilderTests.netcoreapp.cs]: Do these two tests between them hit both of the new OutOfMemoryException you added?

Comment (danmosemsft) [reply src/System.Runtime/tests/System/Text/StringBuilderTests.netcoreapp.cs]: I guess so.

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please

**Comment by jkotas** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please

File: src/System.Runtime/tests/System/Text/StringBuilderTests.netcoreapp.cs [hunk 2]
```diff
@@ -445,5 +445,60 @@ public static void EqualsIgnoresMaxCapacity()
...
+            Assert.True(sb.Length <= sb.Capacity);
+            Assert.True(sb.Capacity <= sb.MaxCapacity);
+        }
+
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsRobustUnderMemoryPressure))]
+        [OuterLoop]
+        public static void Replace_OutOfMemory_DoesNotCorruptState()
+        {
+            var sb = new StringBuilder();
+            sb.Append('a', 1_000_000);
```
Comment (danmosemsft) [inline src/System.Runtime/tests/System/Text/StringBuilderTests.netcoreapp.cs]: Maybe randomize these two lengths  as well?

File: src/System.Runtime/tests/System/Text/StringBuilderTests.netcoreapp.cs [hunk 3]
```diff
@@ -445,5 +445,62 @@ public static void EqualsIgnoresMaxCapacity()
...
+            }
+            catch (OutOfMemoryException)
+            {
+            }
+            catch (ArgumentException)
+            {
+            }
+
+            Assert.True(sb.Length <= sb.Capacity);
+            Assert.True(sb.Capacity <= sb.MaxCapacity);
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/Text/StringBuilderTests.netcoreapp.cs]: Would this be better as:
```C#
Assert.InRange(sb.Capacity, sb.Length, sb.MaxCapacity);
```
?

**Comment by jkotas** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please

**Comment by jkotas** [comment]
@dotnet-bot test Linux x64 Release Build  please

**Pull request closed without merging**